lib/watch/index.js

```diff
@@ -0,0 +1,86 @@
+const path = require('path')
+const cp = require('child_process')
+
+const GRACEFUL_SHUT = 'GRACEFUL SHUTDOWN'
+const FULLY_CLOSED = 'FULLY CLOSED'
+const READY = 'ready'
+
+const EventEmitter = require('events')
+const chokidar = require('chokidar')
+const forkPath = path.join(__dirname, './fork.js')
+
+const emitter = new EventEmitter()
+
+let childs = []
+
+const stop = (watcher, err) => {
+  childs.forEach(function (child) {
+    child.kill()
+  })
+  watcher.close()
+}
+
+const watch = function (args) {
+  let childEvent = null
+  const run = (event) => {
+    const _child = cp.fork(forkPath, args, {
+      env: Object.assign({}, process.env, childEvent),
+      cwd: process.cwd(),
+      encoding: 'utf8'
+    })
+
+    _child.on('exit', function (code, signal) {
+      if (signal === 'SIGUSR2') {
+        childEvent = { childEvent: { type: 'start' } }
+        childs.push(run('restart'))
+      }
+      return null
+    })
+
+    _child.on('message', (childEvent) => {
+      const { type, err } = childEvent
+      if (err) {
+        emitter.emit('error', err)
+        return null
+      }
+
+      if (type === FULLY_CLOSED) {
+        childs.shift().kill('SIGUSR2')
+        return null
+      }
```

Comment:
why do we need to wait for this signal?

Reply:
Only use SIGUSR2 signal kill child process, the main process would fork new child process. other signals don't.

Reply:
I mean, why are waiting for a FULLY_CLOSED message from the child process before restarting?

Reply:
Because the signal would send in fastify.close() callback or hook timing that call in child_process.

Reply:
should child_process killed in itself instead of don't send message parent process?

Reply:
Because new child process should fork in fastify.close() timing and old child process would kill.

Reply:
I think we should just `process.exit(0)`  in the child process within the `fastify.close()` callback. You should not be waiting for a message.